src/PointwiseFunctions/AnalyticSolutions/Errors.hpp

```diff
@@ -0,0 +1,73 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+
+#include "DataStructures/DataBox/Tag.hpp"
+#include "DataStructures/Variables.hpp"
+#include "DataStructures/VariablesTag.hpp"
+#include "Domain/Tags.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/Tags.hpp"
+#include "Time/Tags.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace Tags {
+/*!
+ * \brief Prefix indicating the difference between two values, such as betwen a
+ * numeric quantity and the corresponding quantity computed from an analytic
+ * solution.
+ */
+template <typename Tag>
+struct Error : db::PrefixTag, db::SimpleTag {
+  using type = typename Tag::type;
+  using tag = Tag;
+};
+
+// @{
+/*!
+ * \brief For each `Tag` in `TagsList`, compute its difference from the
+ * analytic solution.
+ */
+template <size_t VolumeDim, typename AnalyticSolutionTag, typename TagsList>
+struct ErrorsCompute;
+
+template <size_t VolumeDim, typename AnalyticSolutionTag, typename... FieldTags>
+struct ErrorsCompute<VolumeDim, AnalyticSolutionTag, tmpl::list<FieldTags...>>
+    : db::add_tag_prefix<::Tags::Error,
+                         ::Tags::Variables<tmpl::list<FieldTags...>>>,
+      db::ComputeTag {
+  using tags_list = tmpl::list<FieldTags...>;
+  using base = db::add_tag_prefix<::Tags::Error, ::Tags::Variables<tags_list>>;
+  using return_type = tmpl::type_from<base>;
+
+  using argument_tags = tmpl::append<
+      tmpl::list<AnalyticSolutionTag,
+                 domain::Tags::Coordinates<VolumeDim, Frame::Inertial>,
+                 Tags::Time>,
+      tags_list>;
+
+  static constexpr void function(
+      const gsl::not_null<return_type*> errors,
+      const db::const_item_type<AnalyticSolutionTag>&
+          analytic_solution_computer,
+      const tnsr::I<DataVector, VolumeDim, Frame::Inertial>& inertial_coords,
+      const double& time, const typename FieldTags::type&... fields) noexcept {
```

Comment:
- take `double` by value
- `typename FieldTags::type` -> `db::const_item_type<FieldTags>`

Reply:
I’m fine with these suggestions, but I wonder how I know when to use db::const_item_type or db::item_type vs Tag::type?

Reply:
Here's my understanding (others can confirm/correct):

- If you know the type explicitly, just write it. E.g. where you know an argument is a `Mesh` or `Scalar<DataVector>`, write that instead of getting the type from some tag.
- If you are dealing with DataBox tags and you're sure they are not base tags, use `db::const_item_type` where the tag is being retrieved from the box and `db::item_type` where it's being mutated.
- If the tags can be base tags, either use `db::[const_]item_type` if you have access to the full DataBox tags list and pass it as second template parameter, or template the function on the argument types and have the compiler infer the types when the function is called.
- If you're sure you're not dealing with a DataBox, or you are sure the tags are simple tags and don't store pointers, you can use `Tag::type`. Prefer `tmpl::type_from<Tag>` over `typename Tag::type` because it's annoying to write `typename`.

Reply:
Thanks @nilsleiffischer! After discussing with @nilsdeppe, I realize I can avoid getting the type at all by just templating function() on the AnalyticSolution and on the field types...I'll do that in my next fixup

Reply:
Right, that's the second part of point 3 above. Not sure if (or why) that's preferable to getting the types by `db::const_item_type` when you know they are not base tags. @nilsdeppe would you say it is? Could you confirm or correct my list above?